opus / vorbis (cleaned)

philippe44 opened the pull request.

philippe44 committed:
add opus + tweak vorbis

cleanup

cleanup

ralph-irving commented:
I've merged the pull request as 2 commits. one for vorbis the other for opus.  You need to define OPUS to include the decoder.
I had to move ops in the capabilites string before flc instead of after.  LMS would always transcode opus files to flac otherwise.  https://github.com/ralph-irving/squeezelite/blob/master/decode.c#L174
I'm experience "breaks" or skips with the opus decoder with 2 cds that I converted from flac to opus. I can make them available for you to download to test if needed.  Also, occasionally when I quit squeezelite, and only after having used the opus decoder I get a double free error.

[13:41:46.003223] output_thread:804 disabling output
^C[13:41:47.355399] slimproto_stop:966 slimproto stop
[13:41:47.455534] decode_close:218 close decode
[13:41:47.458203] stream_close:424 close stream
*** glibc detected *** ./squeezelite: double free or corruption (out): 0xb701d008 ***
======= Backtrace: =========
/lib/i386-linux-gnu/i686/cmov/libc.so.6(+0x6ff41)[0xb728ef41]
/lib/i386-linux-gnu/i686/cmov/libc.so.6(+0x717a8)[0xb72907a8]
/lib/i386-linux-gnu/i686/cmov/libc.so.6(cfree+0x6d)[0xb72938ed]
./squeezelite[0x8055e28]
/lib/i386-linux-gnu/i686/cmov/libc.so.6(__libc_start_main+0xe6)[0xb7235e46]
./squeezelite[0x8052bad]

Unfortunately, when I run squeezelite under gdb I've not been able to get a backtrace or even the double free error message.

philippe44 committed:
opus_read wants buf_size in "values" which means u16 ...

philippe44 committed:
solve merge

philippe44 commented:
I saw the problem and corrected it. I misundertood the definition of "values" in the opus_read API and that lead to memory overwrite.
Re where is is OPUS in the LMS order, what surprises me is that LMS does not seem to try to find a direct match before trying transcoding rules. Don't you think we should solve that (or is this a convert.conf order problem)

ralph-irving commented:
I've listened to 6 opus tracks so far and there have been no artifacts in the audio.  I've merged the fix for opus_read.  Thanks!
Re LMS transcoding rules.  It would be worth looking into.  However, I wonder what impact changing that behaviour might have on plugins that are designed within the current restriction?  C3PO in particular.

philippe44 committed:
remove compile warning

philippe44 committed:
cleanup

philippe44 committed:
remove redundant "opened" flag

philippe44 commented:
I was wondering if there are still some people complaining about vorbis "early" decode termination (we got that on spotty when Michael enabled direct ogg streaming). Could this be due to the fact that ogg (and opus) return DECODE_COMPLETE when no frame is returned from the decoder. 

I'm not sure that these decoder might not return nothing when they need more than what is given to them (in our case, streambuf does not have enough). So I would rather continue and declare completion only when streambuf is empty and stream state is disconnected. It will cause maybe one extra call to decode() but it really does not matter.

Or have the early decode problem been solved by changing min_read_size?

philippe44 committed:
no frames is not end of stream

philippe44 committed:
Merge branch 'master' into master

ralph-irving commented:
I still have noticed spotty prematurely skipping to the next track maybe once a week.  Thanks for making the vorbis change like you did for opus.  I committed these two changes separately.  Looking at the tremor decoder in squeezeplay, the devs added an  OGG_STARVED status to handle the no frames is not end of stream as well.  I hadn't noticed that before.

https://github.com/ralph-irving/squeezeplay/blob/master/src/Tremor/ogg.h#L211

https://github.com/ralph-irving/squeezeplay/blob/master/src/squeezeplay/src/audio/decode/decode_vorbis.c#L49

philippe44 committed:
decode non ending

In vorbis (maybe opus) it might happen that there are still some frames
in the buffer but the decoder does not decode anything anyhow and stops
progressing. So, when no frames is decoded and streaming has ended,
still declare end of decode

philippe44 committed:
CRLF ...

philippe44 committed:
conflicts ...

philippe44 committed:
more conflicts

philippe44 committed:
merge

philippe44 commented:
Where did you find the "official" ubuntu/debian alac development package?

ralph-irving commented:
No. I haven't.  I tried Mike Brady's alac https://github.com/mikebrady/alac since it has autoconfig, but it crashes squeezelite armv6 but your repo works fine with it.  Both repositories currently crash my win32 builds as soon as I try to play an alac track.

philippe44 commented:
Just to make sure I get that right, alac does not work at all on win32? Where did you get the alac decoder itself? On my system, my win32 build works well

ralph-irving commented:
Yes. It crashes right after the alac_decode setting track_start. I'm using your alac repo and created a vs2008 project to create the static lib.  I haven't had a chance to run it in debug to see what's happening yet.

ralph-irving commented:
So I found https://github.com/TimothyGu/alac which had VS2008 and debian package projects.  The static alac lib created using this works on windows.  I haven't looked at the decoder code differences but this repo generates a lot less warnings with vs2008.
